src/lua/protobuf.lua

```diff
@@ -0,0 +1,616 @@
+        error(('Input data for %q field is %q and do not fit in %q')
+            :format(field_def.name, value, field_def.type))
+    end
+end
+
+validate_scalar = function(field_def, value)
+    local scalar_def = scalars[field_def.type]
+    assert(scalar_def.accept_type ~= nil)
+    validate_type(field_def, value, scalar_def.accept_type)
+    if scalar_def.accept_type == 'string' then
+        validate_length(value)
+    end
+    if type(value) == 'number' and is_nan(value) then
+        error(('Input data for %q field is NaN'):format(field_def.name))
+    end
+    if type(value) == 'number' and is_inf(value) then
+        error(('Input data for %q field is inf'):format(field_def.name))
+    end
+    if scalar_def.integral_only then
+        check_integer(field_def, value)
+    end
+    if scalar_def.limits ~= nil then
+        local value_type = type(value)
+        if value_type == 'cdata' then
+            value_type = ffi.istype(int64_t, value) and 'int64'
+                or 'uint64'
+        end
+        validate_range(field_def, value, scalar_def.limits[value_type])
+    end
+end
+
+-- }}} Validation
+
+
+-- {{{ Scalar type definitions
+
+scalars.float = {
+    accept_type = 'number',
+    encode_as_packed = true,
+    integral_only = false,
+    limits = {
+        number = {-0x1.fffffep+127, 0x1.fffffep+127},
+    },
+    encode = wireformat.encode_float,
+}
+
+scalars.fixed32 = {
+    accept_type = {'number', 'cdata'},
+    encode_as_packed = true,
+    integral_only = true,
+    limits = {
+        number = {0, 2^32 - 1},
+        int64 = {0LL, 2LL^32 - 1},
+        uint64 = {nil, 2ULL^32 - 1},
+    },
+    encode = wireformat.encode_fixed32,
+}
+
+scalars.sfixed32 = {
+    accept_type = {'number', 'cdata'},
+    encode_as_packed = true,
+    integral_only = true,
+    limits = {
+        number = {-2^31, 2^31 - 1},
+        int64 = {-2LL^31, 2LL^31 - 1},
+        uint64 = {nil, 2ULL^31 - 1},
+    },
+    encode = wireformat.encode_fixed32,
+}
+
+scalars.double = {
+    accept_type = 'number',
+    encode_as_packed = true,
+    integral_only = false,
+    encode = wireformat.encode_double,
+}
+
+scalars.fixed64 = {
+    accept_type = {'number', 'cdata'},
+    encode_as_packed = true,
+    integral_only = true,
+    limits = {
+        number = {0, 2^64 - 1},
+        int64 = {0LL, nil},
+    },
+    encode = wireformat.encode_fixed64,
+}
+
+scalars.sfixed64 = {
+    accept_type = {'number', 'cdata'},
+    encode_as_packed = true,
+    integral_only = true,
+    limits = {
+        number = {-2^63, 2^63 - 1},
+        uint64 = {nil, 2ULL^63 - 1},
+    },
+    encode = wireformat.encode_fixed64,
+}
+
+scalars.string = {
+    accept_type = 'string',
+    encode_as_packed = false,
+    encode = wireformat.encode_len,
+}
+
+scalars.bytes = scalars.string
+
+scalars.int32 = {
+    accept_type = {'number', 'cdata'},
+    encode_as_packed = true,
+    integral_only = true,
+    limits = {
+        number = {-2^31, 2^31 - 1},
```

Comment:
I'm not sure that using constants is a nice idea. But please check my diff and if it is okay I will implement this

```lua
 local RESERVED_FIELD_ID_MAX = 19999
 local MAX_FIELD_ID = 2^29 - 1

+-- Number limits for int32 and int64
+local MAX_FLOAT = 0x1.fffffep+127
+local MAX_INT32 = 2^32 - 1
+local MAX_INT64 = 2^64 - 1
+local MIN_SINT32 = -2^31
+local MAX_SINT32 = 2^31 - 1
+local MIN_SINT64 = -2^63
+local MAX_SINT64 = 2^63 - 1
+
+-- Cdata limits for int32_t and int64_t
+local MAX_INT32_LL = 2LL^32 - 1
+local MAX_INT32_ULL = 2ULL^32 - 1
+local MAX_INT64_LL = 2LL^64 - 1
+local MAX_INT64_ULL = 2ULL^64 - 1
+local MIN_SINT32_LL = -2LL^31
+local MAX_SINT32_LL = 2LL^31 - 1
+local MAX_SINT32_ULL = 2ULL^31 - 1
+local MIN_SINT64_LL = -2LL^63
+local MAX_SINT64_LL = 2LL^63 - 1
+local MAX_SINT64_ULL = 2ULL^63 - 1
+
+
@@ -451,9 +481,9 @@ scalars.uint32 = {
     encode_as_packed = true,
     integral_only = true,
     limits = {
-        number = {0, 2^32 - 1},
-        int64 = {0LL, 2LL^32 - 1},
-        uint64 = {nil, 2ULL^32 - 1},
+        number = {0, MAX_INT32},
+        int64 = {0LL, MAX_INT32_LL},
+        uint64 = {nil, MAX_INT32_ULL},
     },
     encode = wireformat.encode_int,
 }
@@ -463,8 +493,8 @@ scalars.int64 = {
     encode_as_packed = true,
     integral_only = true,
     limits = {
-        number = {-2^63, 2^63 - 1},
-        uint64 = {nil, 2ULL^63 - 1},
+        number = {MIN_SINT64, MAX_SINT64},
+        uint64 = {nil, MAX_SINT64_ULL},
     },
     encode = wireformat.encode_int,
 }
```

Reply:
I would suggest using MAX_UINT instead of MAX_INT. Also, I believe that `MAX_INT64_LL`, `MAX_INT64_ULL`, `MAX_SINT64_LL` and `MIN_SINT64_LL` don't really make sense, do they? Also, I didn't think yesterday, but `2^63 - 1` is actually equal to `2^63` (since it is `double`). Due to this `MAX_INT64` is actually should be `2^63 - 1024` (not sure though, better to check it).

Reply:
I agree that `MAX_INT64_LL`, `MAX_INT64_ULL`, `MAX_SINT64_LL`, `MIN_SINT64_LL` are useless - deleted them. Changed `MAX_INT` to `MAX_UINT`.

According to my tests numbers from [2^63-512, 2^63+1024] convert into 2^63. As 2^63 is out of range for SINT64 I suppose that `MAX_SINT64` needs to be 2^63-513 or `0x7ffffffffffffc00`.

For the low limit -2^63 is acceptable to encode so the current limit is acceptable but I also will convert it to hex: `0x8000000000000000`.

As for UINT64 the numbers from [2^64-1024, 2^64+2048] convert into 2^64 which again is out of range of UINT64 so the correct limit will be 2^64-1025 or 0xfffffffffffff800. 

```lua
@@ -9,6 +9,24 @@ local RESERVED_FIELD_ID_MIN = 19000
 local RESERVED_FIELD_ID_MAX = 19999
 local MAX_FIELD_ID = 2^29 - 1

+-- Number limits for int32 and int64
+local MAX_FLOAT = 0x1.fffffep+127
+local MAX_UINT32 = 2^32 - 1
+local MAX_UINT64 = 0xfffffffffffff800
+local MIN_SINT32 = -2^31
+local MAX_SINT32 = 2^31 - 1
+local MIN_SINT64 = 0x8000000000000000
+local MAX_SINT64 = 0x7ffffffffffffc00
+
+-- Cdata limits for int32_t and int64_t
+local MAX_UINT32_LL = 2LL^32 - 1
+local MAX_UINT32_ULL = 2ULL^32 - 1
+local MIN_SINT32_LL = -2LL^31
+local MAX_SINT32_LL = 2LL^31 - 1
+local MAX_SINT32_ULL = 2ULL^31 - 1
+local MAX_SINT64_ULL = 2ULL^63 - 1
+
+
```

Reply:
It looks good enough for me.

P.S. I will resolve these comments when I see fixes on the branch.